Native reverse pages over RPC

paszkow opened the pull request:
Drop half-reversed (legacy) format of query::partition_slice.

The select query builds a fully reversed (native) slice for reversed queries and use it together with a reversed
schema to construct query::read_command that is further propagated to the database.

A cluster feature is added to support nodes that still operate on half-reversed slices. When the feature is turned off:
- query::read_command is transformed (to have table schema and half-reversed slices) before sending to other nodes
- query::read_command is transformed (to have query schema (reversed) and reversed slices) after receiving it from other nodes
- Similarly, mutations are transformed. They are reversed before being sent to other nodes or after receiving them from other nodes.

Additional manual tests were performed to test a mixed-node cluster:

1. 3-node cluster with one node upgraded: reverse read queries performed on an old node
2. 3-node cluster with one node upgraded: reverse read queries performed on a new node
3. 3-node cluster with one node upgraded and all its sstable files deleted to trigger repair: reverse read queries performed on an old node
4. 3-node cluster with one node upgraded and all its sstable files deleted to trigger repair: reverse read queries performed on a new node

All reverse read queries above consists of:

- single-partition reverse reads with no clustering key restrictions, with single column restrictions and multi column restrictions both with and without paging turned on
- multi-partition reverse reads with range restrictions with optional partition limit and partial ordering

The exact same tests were also performed on a fully upgraded cluster.

Fixes https://github.com/scylladb/scylladb/issues/12557

bhalevy commented on db/config.cc:
```diff
@@ -1066,8 +1066,6 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , reversed_reads_auto_bypass_cache(this, "reversed_reads_auto_bypass_cache", liveness::LiveUpdate, value_status::Used, false,
             "Bypass in-memory data cache (the row cache) when performing reversed queries.")
-    , enable_optimized_reversed_reads(this, "enable_optimized_reversed_reads", liveness::LiveUpdate, value_status::Used, true,
-            "Use a new optimized algorithm for performing reversed reads.")
```
@mykaul can we verify that this option is indeed unused out in the field?

mykaul replied:
I'm not sure we have the facility (right now). We could if we make an effort - but does it matter - what's the impact of removing it?

paszkow replied:
It is enabled by default. When set to false, however, the reversed queries are performed as forward ones and wrapped into a reversing reader. According to my conversation with @denesb, it was the very first implementation of the reversed queries. It was kept as a back door option in case the reverse queries implementation does not work properly.

mykaul replied:
If we haven't heard of anyone complaining and thus using the fallback, I think it's safe to drop it. (can we include a deprecation warning in 6.0 release notes - @tzach ?)

bhalevy replied:
> It is enabled by default. When set to false, however, the reversed queries are performed as forward ones and wrapped into a reversing reader. According to my conversation with @denesb, it was the very first implementation of the reversed queries. It was kept as a back door option in case the reverse queries implementation does not work properly.

So if we find issues in the field with native reverse queries as per this PR,
will there be a way to fallback to the current implementation (before this PR)?

avikivity replied:
However, given there is a cluster feature for it, it's easy to add configuration too.

avikivity replied:
And for the original question - I'm sure no one touched this configuration.

bhalevy commented on replica/table.cc:
```diff
@@ -208,24 +208,14 @@ flat_mutation_reader_v2
                            mutation_reader::forwarding fwd_mr) const {
     if (_virtual_reader) [[unlikely]] {
-        return (*_virtual_reader).make_reader_v2(s, std::move(permit), range, query_slice, trace_state, fwd, fwd_mr);
+        return (*_virtual_reader).make_reader_v2(s, std::move(permit), range, slice, trace_state, fwd, fwd_mr);
```
nit: this cosmetic change is not required, at least not in this patch, and it hurts reviewability.

avikivity commented:
@yaronkaikov what's with CI?

avikivity commented on replica/table.cc:
```diff
@@ -253,7 +243,7 @@ table::make_reader_v2(schema_ptr s,
 
     const auto bypass_cache = slice.options.contains(query::partition_slice::option::bypass_cache);
-    if (cache_enabled() && !bypass_cache && !(reversed && _config.reversed_reads_auto_bypass_cache())) {
+    if (cache_enabled() && !bypass_cache && !(slice.is_reversed() && _config.reversed_reads_auto_bypass_cache())) {
```
I wonder why we have reversed_reads_auto_bypass_cache (maybe due to earlier cache weakness in this area?). We can consider removing it too.

denesb replied:
We had native reverse implementation for sstables, before we had it for cache. So we had this option to bypass the cache, which at the time would still use the old reverse-whole-partition approach.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -375,8 +379,8 @@ select_statement::do_execute(query_processor& qp,
-            _schema->id(),
-            _schema->version(),
+            _query_schema->id(),
+            _query_schema->version(),
```
What's the compatibility impact? read_command is passed on the wire. What happens if we talk to an old node?

paszkow replied:
If we have a cluster with an old node, then the read command is transformed before it is sent: a table schema is sent and the slice in the legacy-reverse form. The transformation is performed in `storege_proxy::remote::send_read_data`

Similarly, when a command is received from an old node. It is transformed into native reverse before data is read.

avikivity replied:
IIRC reversed schema is transformed into a normal schema before pulling (based more on guessing how it works than real memory).

avikivity commented on utils/value_or_reference.hh:
```diff
@@ -0,0 +1,25 @@
+ */
+
+#pragma once
+
```
Add: #include <optional> so it's self-contained.

avikivity commented on multishard_mutation_query.cc:
```diff
@@ -1001,13 +1001,13 @@ future<std::tuple<foreign_ptr<lw_shared_ptr<reconcilable_result>>, cache_tempera
         tracing::trace_state_ptr trace_state,
         db::timeout_clock::time_point timeout) {
-    schema_ptr query_schema = cmd.slice.is_reversed() ? table_schema->make_reversed() : table_schema;
+    schema_ptr table_schema = cmd.slice.is_reversed() ? query_schema->make_reversed() : query_schema;
```
@denesb the notion of a query schema also applies to mutation fragment selection. It has the same keys as the base table but the columns of the mutation_fragment transformation. I don't think we should do anything now but keep it in mind if it helps one day.

denesb replied:
Same here (consider using `get_reversed()` instead of `make_reversed()`)

avikivity commented on mutation/mutation_partition.cc:
```diff
@@ -2251,24 +2247,11 @@ to_data_query_result(const reconcilable_result& r, schema_ptr s, const query::pa
+    for (const partition& p : r.partitions()) {
+        const auto res = co_await p.mut().consume_gently(s, adaptor);
+        if (res.stop == stop_iteration::yes) {
+            break;
```
I don't understand how you can drop this. If the data came from an old node, it would be a half-reversed mutation, no?

paszkow replied:
Yes it is but same as for the `read_command` I reverse mutations before they are applied. Thus I do not need to distinguished between versions. When applied, it will already be in the native reverse.

avikivity replied:
I don't understand the answer.

paszkow replied:
When the data comes from an old node it is first reversed in `abstract_read_executor::make_mutation_data_request` https://github.com/scylladb/scylladb/blob/b9f7e2f45dd4d20695f1f5f94fe68e42868c80d6/service/storage_proxy.cc#L5039
So I can treat it as it came from a new node.

avikivity replied:
Very good. The patch structure now makes it visible to a diligent reader (though I admit I'm not diligent enough here).

avikivity commented on replica/table.cc:
```diff
@@ -2994,10 +2994,7 @@ table::mutation_query(schema_ptr s,
     std::exception_ptr ex;
   try {
-    // Un-reverse the schema sent to the coordinator, it expects the
-    // legacy format.
```
It may still expect it!

paszkow replied:
Agree. When the coordinator is an old node the data is reversed first before sending it back: `storage_proxy::remote::handle_read` https://github.com/scylladb/scylladb/blob/b9f7e2f45dd4d20695f1f5f94fe68e42868c80d6/service/storage_proxy.cc#L819

avikivity replied:
Aha

avikivity commented on test/boost/flat_mutation_reader_test.cc:
```diff
@@ -1129,16 +1129,14 @@ SEASTAR_THREAD_TEST_CASE(test_reverse_reader_v2_is_mutation_source) {
                 streamed_mutation::forwarding fwd_sm,
                 mutation_reader::forwarding fwd_mr) mutable {
             flat_mutation_reader_v2 rd(nullptr);
```
> unit tests: fix tests related to reversed queries

I don't understand what you're fixing.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -210,6 +210,36 @@ class storage_proxy::remote {
+
+        return result;
+    }
+
```
This patch should come earlier in the series. When you remove support for half-reversed mutations in one place, insert a call to one of these at its call site, thus showing how you maintain compatibility, rather than removing compatibility and re-adding it later in separate places.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -210,6 +210,36 @@ class storage_proxy::remote {
+
+    // Transform a reconcilable_result by reversing stored mutations. The transformation is done only for reverse read queries
+    // in a mixed node cluster when the native_reverse_queries feature flag is turned off.
+    reconcilable_result transform_to_from_legacy_format_for_native_reversed_queries(reconcilable_result result, bool reversed)
```
reconcilable_result&& to avoid copies

avikivity commented on service/storage_proxy.cc:
```diff
@@ -210,6 +210,36 @@ class storage_proxy::remote {
 
+    // Transform a read command by reversing the schema version and the bounds of the clustering key ranges. The transformation
+    // is done only for reverse read queries in a mixed node cluster when the native_reverse_queries feature flag is turned off.
+    query::read_command transform_to_from_legacy_format_for_native_reversed_queries(query::read_command cmd, bool reversed)
```
transform_to_from? confusing name

avikivity commented on service/storage_proxy.cc:
```diff
@@ -210,6 +210,36 @@ class storage_proxy::remote {
+    reconcilable_result transform_to_from_legacy_format_for_native_reversed_queries(reconcilable_result result, bool reversed)
+    {
+        if (!reversed || _sp.features().native_reverse_queries)
+            return result;
```
_sp.feature checks should only be done on the client side. On the server side, only check the command.

Otherwise you are subject to the following race:

client: advertise feature
server: advertise feature
client: recognize cluster has feature
client: send request that relies on the feature
server: mishandle request
server: recognize cluster has feature

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -392,6 +392,8 @@ select_statement::do_execute(query_processor& qp,
             options.get_timestamp(state));
     command->allow_limit = db::allow_per_partition_rate_limit::yes;
+    tracing::trace(state.get_trace_state(), "Executing read query (reversed {}): table schema {}, query schema {}",
+        slice.is_reversed(), _schema->version(), _query_schema->version());
```
Most of these aren't useful for tracing. They can be useful for debugging, I think logging with trace level is more appropriate.

I'm sure we have a trace at the beginning of a query, we can add reversed there.

avikivity commented:
Looks good and well done tackling a difficult area.

My main comment is about the series structure - first breaking compatibility and then re-introducing it. It reduces confidence in upgrades.

@michoecho please review the cache/mutation_partition related areas.

How are reversed queries during upgrade covered? I'm sure we added some coverage during the Great Reversing of 2021. @roydahan

denesb commented on replica/memtable.cc:
```diff
@@ -450,7 +451,7 @@ class scanning_reader final : public flat_mutation_reader_v2::impl, private iter
          , _slice(slice)
          , _fwd_mr(fwd_mr)
-     { }
+     {}
```
Avoid unnecessary changes like this.

denesb reviewed (changes requested):
The structure of this PR needs to change, such that at each commit, the code compiles and the tests are passing.
So you first introduce the feature flag and the transformation code in storage proxy, and then introduce the changes such that at each step you still have a functioning scylla.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9229/)
#### Build Details:
- Duration: 1 hr 38 min
- Builder: spider1.cloudius-systems.com

avikivity commented:
> The title ("Implement native reverse queries") is misleading, native reverse queries are already implemented, this just drops the legacy path.

That's not accurate either. This is about having reversed pages over RPC. Everywhere else we already had native reversed queries. And we don't drop the legacy path, that comes in two years.

roydahan commented:
> How are reversed queries during upgrade covered? I'm sure we added some coverage during the Great Reversing of 2021. @roydahan

It’s a good point, I don’t think we have it during upgrades.
@yarongilor haven’t you added such a workload?
I recall a lot back and forth on the reverse query PRs.

roydahan commented:
I guess you can add one to the workload or show @lukepio how to add and test it.

michoecho commented on service/storage_proxy.cc:
```diff
@@ -4551,7 +4551,7 @@ class data_read_resolver : public abstract_read_resolver {
         bool always_return_static_content = cmd.slice.options.contains<query::partition_slice::option::always_return_static_content>();
-        mp.compact_for_query(s, m.decorated_key(), cmd.timestamp, ranges, always_return_static_content, is_reversed, limit);
+        mp.compact_for_query(s, m.decorated_key(), cmd.timestamp, ranges, always_return_static_content, false, limit);
         return primary_key{m.decorated_key(), get_last_reconciled_row(s, mp, is_reversed)};
```
Does reconciliation actually work after this PR? If everything read-related is "native", isn't the reconciliation result going to be reversed, and shouldn't we be unreversing it somewhere before applying it to replicas?

What tests did you run for this PR?

paszkow replied:
I run the following tests

1. Start a 2-node cluster, create a table with RF=2 and populate it with data
2. Stop one of the nodes and delete all its sstable files.
3. Restart the node
4. Execute a reverse read on either of the nodes with consistency level set to ALL
5. Execute a reverse/forward read with consistency level set to ONE on the node that was stopped

Then I altered my test case to not just delete all files but rather introduce range tombstones (also overlapping) by executing multiple `DELETE` statements on the running node. Similarly with `UPDATE` statements.

Both test cases were testing with paging turned on and off. Such a test is missing in dtest so it should end up there to validate reconciliation still works.

avikivity replied:
Don't we have unit tests for reconciliation?

michoecho commented on readers/mutation_readers.cc:
```diff
@@ -1119,12 +1119,11 @@ std::deque<mutation_fragment_v2> reverse_fragments(const schema& schema, reader_
 make_flat_mutation_reader_from_fragments(schema_ptr schema, reader_permit permit, std::deque<mutation_fragment_v2> fragments,
-        const dht::partition_range& pr, const query::partition_slice& query_slice) {
-    const auto reversed = query_slice.is_reversed();
+        const dht::partition_range& pr, const query::partition_slice& slice) {
```
This patch changes the functions, but doesn't adjust any callers. Are there no callers which pass reverse slices here? Or does the adjustment happen in a different commit?

michoecho commented on query-request.hh:
```diff
@@ -148,13 +147,6 @@ constexpr auto max_rows_if_set = std::numeric_limits<uint32_t>::max();
-// format:
-// * legacy format
-// * native format
-// The wire format uses the legacy format. See docs/dev/reverse-reads.md
```
`docs/dev/reverse-reads.md` should be adjusted accordingly.

michoecho reviewed (changes requested):
I agree with comments left by others, and I left some of my own.

Most importantly: something is telling me that the reconciliation path is doing something weird after the changes. Is it tested properly?

bhalevy commented on replica/database.cc:
```diff
@@ -2099,7 +2094,7 @@ future<> database::apply_hint(schema_ptr s, const frozen_mutation& m, tracing::t
         dblog.trace("apply hint {}", m.pretty_printer(s));
     }
-    if (!s->is_synced()) {
+    if (!(s->is_synced() || s->get_reversed()->is_synced())) {
```
Also, it deserves an explanation how is_synced can be different on the schema and its reversed schema.

Maybe the right direction is to make sure they are in sync with each other if that's correct, and then you won't need this change.

If we do, then better print also the reversed schema version in the error message since if we get the reversed table schema here, I'm not sure its (reversed) version would be helpful, as far as finding it in the logs.

bhalevy commented on replica/memtable.cc:
```diff
@@ -730,12 +727,9 @@ memtable::make_flat_reader_opt(schema_ptr s,
-
+        auto cr = query::clustering_key_filter_ranges::get_native_ranges(*s, slice, dk.key());
         bool digest_requested = slice.options.contains<query::partition_slice::option::with_digest>();
+
```
stray change

bhalevy commented on sstables/kl/reader.cc:
```diff
@@ -1154,19 +1156,20 @@ class sstable_mutation_reader : public mp_row_consumer_reader_k_l {
             : mp_row_consumer_reader_k_l(std::move(schema), permit, std::move(sst))
-            , _consumer(this, _schema, std::move(permit), slice, std::move(trace_state), fwd, _sst)
+            , _slice_holder(std::move(slice))
+            , _slice(_slice_holder.get())
```
Alternatively maybe the _slice member should be eliminated and  be replaced with a slice() getter that will return _slice_holder().get()

bhalevy replied:
If its only use is the line below, then it makes even a stronger case for dropping _slice

bhalevy commented on utils/value_or_reference.hh:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
```
We're in 2024 :)

bhalevy commented on utils/value_or_reference.hh:
```diff
@@ -0,0 +1,25 @@
+    std::optional<T> _opt;
+    const T& _ref;
+
+    value_or_reference(T&& v) : _opt(std::move(v)), _ref(*_opt) {}
```
Agreed.
Also, if we ever need it to be default constructed we may want to keep a pointer rather than a refernce member and initialize it to nullptr.

paszkow replied:
It was extracted from `sstables/mx/reader.cc` exactly as it was written. The goal was to unify the interfaces between readers. Got into a segfault because I did not expect a difference. Adding `explicit` will also increase readability so definitely to be added.

paszkow commented:
In v2:

- Rearranged commits to ensure scylla is not broken at any step
- Eliminated a potential server-client race
- Dropped config's reversed_reads_auto_bypass_cache flag
- Additional cleanup the unused code:
      - `mutation_partition::compact_for_query drop` reverse parameters
      - clustering_key_filter: unify get_ranges and get_native_ranges
      - streamed_mutation_freezer: drop reverse parameter

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -274,13 +277,16 @@ select_statement::make_partition_slice(const query_options& options) const
     if (_is_reversed) {
         std::reverse(bounds.begin(), bounds.end());
+        for (auto& bound : bounds)
+            bound = query::reverse(bound);
```
Missing curly braces

avikivity replied:
This patch is too difficult for me to review. It touches too many places, so I can't tell if somewhere is missed or if some place does something wrong.

One approach is to use a wave to make those changes. Each patch adjusts a callee and its callers, or a caller and its callees, leaving the rest of the system untouched. Eventually the wave propagates over the whole code base. In fact that's what we used during the original native reversed queries transformation.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9514/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9514/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/counter_test
:small_blue_diamond: boost/flat_mutation_reader_test
:small_blue_diamond: boost/multishard_mutation_query_test
:small_blue_diamond: boost/mutation_query_test
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/row_cache_test
:small_blue_diamond: boost/sstable_conforms_to_mutation_source_test
#### Failed Tests (100/347):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.10.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.100.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.11.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.12.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.13.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.14.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.15.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.16.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9514/testReport/junit/boost.mutation_test.test_trim_rows.17.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 55 min
- Builder: spider7.cloudius-systems.com

denesb commented on service/storage_proxy.cc:
```diff
@@ -191,6 +191,27 @@ template <> struct fmt::formatter<service::storage_proxy_remote_read_verb> : fmt
+    cmd->schema_version = schema->version();
+    cmd->slice = query::legacy_reverse_slice_to_native_reverse_slice(*schema, cmd->slice);
+
+    return cmd;
```
You need `return std::move(cmd)` here to move `cmd`, because unintuitively, a `&&` which has a name, is not a candidate for move.

bhalevy replied:
so why pass `cmd` by rvalue ref?  can be passed by value just as well.

paszkow replied:
It was initially passed by value and I was asked to make it rvalue. That's a minor thing.

denesb replied:
In general, using an rvalue is only needed when you want to force ownership change. It is best to prefer value params in the general case.

denesb commented on service/storage_proxy.cc:
```diff
@@ -191,6 +191,27 @@ template <> struct fmt::formatter<service::storage_proxy_remote_read_verb> : fmt
+        m = frozen_mutation(reverse(m.unfreeze(schema)));
+    }
+
+    return result;
```
Same here.

bhalevy replied:
ditto, can just pass result by value, no?

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
+
+        auto f_r = co_await coroutine::as_future([&]() -> future<bool> {
+            // Check if we have a forward query
+            if (!cmd->slice.is_reversed())
```
Curly brackets.

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
         auto src_ip = src_addr.addr;
         auto timeout = t ? *t : db::no_timeout;
+
+        auto f_r = co_await coroutine::as_future([&]() -> future<bool> {
```
Need a better name for this `f_r`.

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
+                return ::make_ready_future<bool>(false);
+
+            // We deal with a reversed query. Check if read command is coming from an old node.
+            // Table schema is then passed with the read command.
```
This sentence is ambiguous. Need to rephrase to be connected with the previous sentence:
"If coordinator is an old node, it will send the tabls schema in the read command".

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
+            }
+
+            // We deal with a reversed query. Check if read command is coming from an new node.
+            // Reversed table schema is then passed with the read command.
```
Same here: "If coordinator is a new node, it will send the reversed table schema in the read command".

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
+
+            // The table schema is outdated. Still do not know if the read command comes from
+            // and old or a new new. Ask coordinator for a schema.
+            return get_schema_for_read(cmd->schema_version, std::move(src_addr), timeout).then_wrapped([&](auto f){
```
space between `)` and `{`

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
+                }
+
+                return ::make_ready_future<schema_ptr>(f.get());
+            }).then([&cmd](auto schema){
```
space between `)` and `{`

denesb commented on service/storage_proxy.cc:
```diff
@@ -721,8 +742,54 @@ class storage_proxy::remote {
+                return ::make_ready_future<schema_ptr>(f.get());
+            }).then([&cmd](auto schema){
+                // A schema is received from the coordinator. Check if it is a table schema.
+                if (schema->version() != cmd->schema_version) {
```
I don't think this is enough to determine whether the coordinator is old or new, we need the first two checks from this method. Maybe move those into a separate function and invoke them from here as well.

paszkow replied:
You are right! Good catch. My comment above is completely misleading.

denesb reviewed (commented):
I think the code in `handle_read()` needs some polishing and the big patch it is in needs splitting as Avi requested.
LGTM otherwise.

paszkow commented:
> I think the code in `handle_read()` needs some polishing and the big patch it is in needs splitting as Avi requested. LGTM otherwise.

Yes, I am in the process of splitting the big commit the way Avi suggested. But I also need to ensure scylla works with every commit, so each time I need to run tests to check nothing is broken.

denesb commented:
> > I think the code in `handle_read()` needs some polishing and the big patch it is in needs splitting as Avi requested. LGTM otherwise.
> 
> Yes, I am in the process of splitting the big commit the way Avi suggested. But I also need to ensure scylla works with every commit, so each time I need to run tests to check nothing is broken.

Often, if the changes are trivial (e.g. propagating some parameter into the next layer), it is enough to just compile the code, or maybe even just compile the changed source files.
As you gain experience, you will become better at gauging how deep you have to go with ensuring that your commit didn't break scylla. If you make a mistake and it does break in a middle commit, and it is merged that way into master, it is not the end of the world. It happens. That is to say, you can try to reduce the amount of checking you do on each commit, especially if said commit is just shuffling parameters around or changes some signature. There is also a `ninja dev-objects` target, which compiles the source files, but doesn't link them (which can be a big time saver).

paszkow commented:
In v2:

- Split big commit into smaller pieces
- Fix determining a reverse format of the read_command in `storage_proxy::remote::handle_read`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9727/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9727/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/counter_test
:small_blue_diamond: boost/flat_mutation_reader_test
:small_blue_diamond: boost/multishard_mutation_query_test
:small_blue_diamond: boost/mutation_query_test
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/row_cache_test
:small_blue_diamond: boost/sstable_conforms_to_mutation_source_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9727/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10124/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2101/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9727/)
#### Failed Tests (22/32939):
- [test_content_update_value](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/system_tables_test/TestConfigTable/Tests___dtest___test_content_update_value) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_content_update_value)
- [test_optimized_reversed_queries_flag](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/reversed_queries_test/TestOptimizedReversedQueriesFlag/Tests___dtest___test_optimized_reversed_queries_flag) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_optimized_reversed_queries_flag)
- [test_optimized_reversed_queries_flag](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/reversed_queries_test/TestOptimizedReversedQueriesFlag/Tests___dtest_with_topology_changes___test_optimized_reversed_queries_flag) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_optimized_reversed_queries_flag)
- [test_content_update_value](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/system_tables_test/TestConfigTable/Tests___dtest_with_topology_changes___test_content_update_value) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_content_update_value)
- [test_service_levels_upgrade.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/%28root%29/test_raft_service_levels/Tests___Unit_Tests___test_service_levels_upgrade_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_service_levels_upgrade.debug.2)
- [test_incremental_read_repair[row-tombstone].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/%28root%29/test_read_repair/Tests___Unit_Tests___test_incremental_read_repair_row_tombstone__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incremental_read_repair[row-tombstone].debug.1)
- [test_incremental_read_repair[partition-tombstone].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/%28root%29/test_read_repair/Tests___Unit_Tests___test_incremental_read_repair_partition_tombstone__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incremental_read_repair[partition-tombstone].debug.1)
- [test_incremental_read_repair[row-tombstone].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/%28root%29/test_read_repair/Tests___Unit_Tests___test_incremental_read_repair_row_tombstone__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incremental_read_repair[row-tombstone].debug.2)
- [test_incremental_read_repair[partition-tombstone].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/%28root%29/test_read_repair/Tests___Unit_Tests___test_incremental_read_repair_partition_tombstone__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incremental_read_repair[partition-tombstone].debug.2)
- [test_incremental_read_repair[row-tombstone].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/9727/testReport/junit/%28root%29/test_read_repair/Tests___Unit_Tests___test_incremental_read_repair_row_tombstone__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_incremental_read_repair[row-tombstone].debug.3)
#### Build Details:
- Duration: 4 hr 53 min
- Builder: spider8.cloudius-systems.com

avikivity commented:
Looks like relevant tests failed.

avikivity commented on query.cc:
```diff
@@ -60,6 +60,20 @@ std::ostream& operator<<(std::ostream& out, const read_command& r) {
 
+lw_shared_ptr<query::read_command> reversed(lw_shared_ptr<query::read_command>&& cmd)
+{
+    return ::make_lw_shared(reversed(std::move(*cmd)));
```
This is destroying *cmd, which might still be used by the caller. It really should have been lw_shared_ptr<const read_command>, but isn't (for historical reasons).

avikivity commented on mutation_query.cc:
```diff
@@ -76,3 +77,20 @@ auto fmt::formatter<reconcilable_result::printer>::format(
+
+lw_shared_ptr<reconcilable_result> reversed(lw_shared_ptr<reconcilable_result>&& result)
+{
+    return ::make_lw_shared(reversed(std::move(*result)));
```
Same problem: if result is shared, then this destroys it for its sharers.

avikivity commented on mutation_query.cc:
```diff
@@ -76,3 +77,20 @@ auto fmt::formatter<reconcilable_result::printer>::format(
+    {
+        auto& m = partition.mut();
+        auto schema = local_schema_registry().get(m.schema_version());
+        m = frozen_mutation(reverse(m.unfreeze(schema)));
```
This has the potential to stall.

If we have similar O(n) operations on reconcilable_result, then we can say it isn't making it worse, but if other reconcilable_result operations are careful to yield, then so must this.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -723,6 +723,56 @@ class storage_proxy::remote {
+        // Detect whether a transformation from legacy reverse format into native reverse
+        // format is necessary before executing the read_command. That happens when the
+        // native_reverse_queries feature is turned off. A custer has mixed nodes.
+        auto f_format_reverse_required = co_await coroutine::as_future([&]() -> future<bool> {
```
This is adding an allocation in a very hot path. Better to fold the inner coroutine into the outer. Or better, handle the common case (fully updated cluster) inline and throw the edge case into a helper (which can be slower).

avikivity commented on service/storage_proxy.cc:
```diff
@@ -6029,12 +6029,16 @@ storage_proxy::query(schema_ptr s,
 future<result<storage_proxy::coordinator_query_result>>
-storage_proxy::query_result(schema_ptr s,
-    lw_shared_ptr<query::read_command> cmd,
+storage_proxy::query_result(schema_ptr s1,
```
Suggest to rename the parameter query_schema to make it abundantly clear what has been transformed.

avikivity replied:
(in .hh too, and the other callees)

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -864,7 +872,7 @@ select_statement::process_results(foreign_ptr<lw_shared_ptr<query::result>> resu
     if (fast_path) {
         return make_ready_future<shared_ptr<cql_transport::messages::result_message>>(make_shared<cql_transport::messages::result_message::rows>(result(
-            result_generator(_schema, std::move(results), std::move(cmd), _selection, _stats),
+            result_generator(_query_schema, std::move(results), std::move(cmd), _selection, _stats),
```
Is result_generator prepared for query_schema?

avikivity commented on service/storage_proxy.cc:
```diff
@@ -772,6 +772,10 @@ class storage_proxy::remote {
+        const bool format_reverse_required = f_format_reverse_required.get();
+
+        if (format_reverse_required)
+            cmd = reversed(std::move(cmd));
```
Braces even for single-statement control flow.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -809,7 +813,11 @@ class storage_proxy::remote {
+                    {
+                        auto&& [result, hit_rate] = r;
+                        return rpc::tuple{format_reverse_required ? make_foreign(reversed(result.release())) : std::move(result), hit_rate};
+                    });
```
To avoid an allocation in the common case (upgraded cluster), apply the continuation conditionally.

(e.g. `f = query_mutations_locally(); if (reverse required) { f = f.then(...); }`)

avikivity commented on service/storage_proxy.cc:
```diff
@@ -4977,6 +4985,8 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
 
 private:
+    const bool _native_reversed_queries_enabled;
+
```
Very good to capture this flag so it doesn't change under our feet.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5018,11 +5029,14 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
         if (_proxy->is_me(ep)) {
             tracing::trace(_trace_state, "read_mutation_data: querying locally");
-            return _proxy->apply_fence(_proxy->query_mutations_locally(_schema, cmd, _partition_range, timeout, _trace_state), fence, _proxy->my_address());
+            co_return co_await _proxy->apply_fence(_proxy->query_mutations_locally(_schema, cmd, _partition_range, timeout, _trace_state), fence, _proxy->my_address());
```
This continuation->coroutine transform in a hot path has to be deliberate to consider performance. Please avoid it for now, we can consider it later if it doesn't regress performance.

avikivity commented on service/storage_proxy.cc:
```diff
@@ -5018,11 +5029,14 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
+            auto cmd1 = format_reverse_required ? reversed(::make_lw_shared(*cmd)) : cmd;
+            auto&& [result, hit_rate] = co_await _proxy->remote().send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout,
+                _trace_state, *cmd1, _partition_range, fence);
+            co_return rpc::tuple{format_reverse_required ? make_foreign(reversed(result.release())) : std::move(result), hit_rate};
```
And here you can apply the conditional continuation trick too.

avikivity commented on partition_snapshot_reader.hh:
```diff
@@ -223,13 +215,9 @@ private:
         } else {
-            if constexpr (Reversing) {
-                opt_reversed_range = query::reverse(*_current_ck_range);
-            }
```
Could we remove Reversing?

avikivity replied:
I guess not, it still needs to know it's scanning the source backwards.

paszkow replied:
You are right. We cannot.

avikivity commented:
Looks good, my confidence is increased although this is still a scary change.

The combination of a rare query type, rare window of exposure (cluster upgrade) and rare sub-cases (reconciliation) makes for many places for bugs to hide.

Let's plan a dedicated test for this designed to find them: run millions of reversed/unreversed  single-partition and multi-partition queries in unupgraded/upgrading/upgraded cluster, while inducing inconsistency.

paszkow commented:
> Looks like relevant tests failed.

dtests fail because I dropped config flags: `enable_optimized_reversed_reads` and `reversed_reads_auto_bypass_cache` The PR https://github.com/scylladb/scylla-dtest/pull/4459 that removes these tests and dropped flags from the tests has been pushed.

Need to double check the unit tests.

paszkow commented:
in v4:

- Use conditional continuation trick
- Fix missing `std::move` or rvalue causing unit tests fail
- Improved readability by renaming variables passed to functions
- Make the function reversing `reconcilable_result` asynchronous with `maybe_yield` to prevent potential stall

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9825/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9825/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/counter_test
:small_blue_diamond: boost/flat_mutation_reader_test
:small_blue_diamond: boost/multishard_mutation_query_test
:small_blue_diamond: boost/mutation_query_test
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/row_cache_test
:small_blue_diamond: boost/sstable_conforms_to_mutation_source_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9825/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2176/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10199/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9825/)
#### Build Details:
- Duration: 10 hr
- Builder: i-0e8f8c0141b4c74d1 (m5ad.8xlarge)

denesb commented on service/storage_proxy.cc:
```diff
@@ -723,6 +723,55 @@ class storage_proxy::remote {
+        // Check if we have a reversed query
+        if (cmd->slice.is_reversed()) {
+            // Compare the schema version provided by read_command with the table schema replica holds.
+            auto table_schema = p->local_db().find_schema(cmd->cf_id);
```
If this node missed a schema update, this can throw and kill the read, even though the read is supposed to survive unsynced schema, via the `get_schema_for_read()` call below.

paszkow replied:
I managed to simulate such a situation by inserting in `migration_manager::merge_schema_from` an error injection that throws an exception. Then one of my nodes in a 3-node cluster missed the information that a table was created. 
Executing a SELECT query with CONSISTENCY ALL, throws an exception when calling `get_schema_for_read()`. It eventually calls `database::find_schema(<table id>)` and fails. I tested it even with the forward read queries to skip this entire `if-`statement. After digging around, I made it work when deliberately changed `if (use_raft)` into `if (false)` in https://github.com/scylladb/scylladb/blob/a99200931a494bb5dc5b4f86015271a6f78dc724/service/migration_manager.cc#L1223
to **not** use raft when syncing a pulled schema. 
With this slight change, it worked for both forward and reversed queries. It looks like we either have a bug or I created a super special corner case with my error injection.

paszkow replied:
Not completely true. I think I created a super special corner case that not necessarily is valid but yet allowed me to test a situation where schema update does not reach the node.

denesb commented on service/storage_proxy.cc:
```diff
@@ -723,6 +723,55 @@ class storage_proxy::remote {
+        // 1. Versions are the same -> legacy format
+        // 2. schema version is equal to reversed table schema -> native format
+        // 3. table schema is outdated
+        static auto is_read_command_in_reversed_legacy_format = [](const query::read_command& cmd, schema_ptr table_schema) -> std::optional<bool> {
```
space between `[]` and `(`.

denesb commented on cql3/statements/select_statement.cc:
```diff
@@ -60,6 +60,8 @@ namespace cql3 {
 
```
Shouldn't this commit come after all the others, which gradually add support for full native reverse support?

paszkow replied:
Depends. I decided to apply changes from top to bottom. Starting from here and propagating the support further down through storage_proxy, database, table up to the readers. Then this patch comes first.

denesb commented on replica/database.cc:
```diff
@@ -1838,19 +1838,19 @@ future<> database::apply_in_memory(const mutation& m, column_family& cf, db::rp_
     return cf.apply(m, std::move(h), timeout);
```
Under what conditions will we ever apply reversed mutations? Such writes will never come via CQL. Is this to support read repair during reverse reads? I think it is possibly simpler to just reverse the mutations we write to replicas as the last step, to avoid these very rarely used code-path changes to the write-path.

paszkow replied:
Exactly for the read-repair during reverse reads. Reversing mutations just before they are sent, could do the work as well. Need to check it. In fact, we already reverse them when do have a mix-node cluster so it should work. I can reverse them always even for new nodes clusters.

Is this solution preferable over what is done now?

denesb replied:
Yes, I think sparing the write path from reverse-awareness is a worthy enough goal to always reverse just before sending the writes.

denesb replied:
Note that just before applying, the mutation has to be reversed anyway, because you cannot apply a reversed mutation directly.
So it is better to reverse on the coordinator and spare the replicas from this.

avikivity replied:
Good catch. Yet another mixed-cluster testing nightmare. I agree sending only unreversed mutations in the write path is much better.

Perhaps we should unreverse before reconciliation to avoid yet another special-ness. I don't see anything special in reconciliation which wouldn't work with reversed mutations, but it's easy to overlook things.

denesb replied:
I think reconciliation works transparently with reverse reads, the only complication is unreversing mutations to be sent to replicas.

paszkow commented on service/storage_proxy.cc:
```diff
@@ -723,6 +723,55 @@ class storage_proxy::remote {
+                format_reverse_required = *is_read_command_in_reversed_legacy_format(*cmd, table_schema);
+            }
+        }
+        (void)format_reverse_required;
```
> Is this a placeholder for future code which will use this variable?

Yes. Otherwise, compiler raises error for unused variable.

paszkow commented:
In v5:

- Send unreversed mutations to replicas as a last step of reconciliation to avoid changes on the write-path
- Simplify the logic in `storage_proxy_::remote` that detects reverse format
- Cover the case where a schema is not yet present on a node
- Resolve conflicts

paszkow commented:
In v6:

- Pass objects by value (instead of rvalue) to `reversed` functions in `mutation_query.cc `
- Fix captures (stop capturing by reference) in `storage_proxy.cc` that were causing rare (1 out of 30) failures in mixed-node cluster tests
- Rebase and resolve conflicts

avikivity commented:
> Fix captures (stop capturing by reference) in storage_proxy.cc that were causing rare (1 out of 30) failures in mixed-node cluster tests

That's good, it indicates the tests were thoroughly exercising the code.

Please document the manual tests in the cover letter.

paszkow commented:
> Please document the manual tests in the cover letter.

Updated. Finally, submitted also PR to dtest that covers mixed-node cluster tests there as well.

avikivity commented:
I see "Unit Tests Custom" failed even though it is marked as Pending.

denesb commented on service/storage_proxy.cc:
```diff
@@ -723,12 +723,48 @@ class storage_proxy::remote {
+            // Note that with the get_schema_for_read call above with reversed table schema, it pulls
+            // the table schema as well. Thus table_schema is always up-to-date.
+            auto table_schema = p->local_db().find_schema(cmd->cf_id);
+            format_reverse_required = *is_read_command_in_reversed_legacy_format(*cmd, table_schema);
```
`is_read_command_in_reversed_legacy_format()` returns an optional and has a return-path which returns a disegaged-one, yet you undconditionally dereference it here.
Either make it not return an optional, or check the return value before dereferencing. The current code is prone to stepping on SEFAULT (`std::optional::operator*()` doesn't check that the optional is engaged before defererencing).

paszkow commented:
> I see "Unit Tests Custom" failed even though it is marked as Pending.

At least it's not due to failed unit tests. All of them passed. In the logs there is information:

```
13:23:28  + /jenkins/workspace/scylla-master/scylla-ci/scylla-pkg/tools/packaging/dpackager -e GITHUB_TOKEN=**** -- gh pr edit 18864 --remove-label 'status/ci in progress'
13:23:48  error connecting to api.github.com
13:23:48  check your internet connection or https://githubstatus.com/
```

mykaul commented:
> > I see "Unit Tests Custom" failed even though it is marked as Pending.
> 
> At least it's not due to failed unit tests. All of them passed. In the logs there is information:
> 
> ```
> 13:23:28  + /jenkins/workspace/scylla-master/scylla-ci/scylla-pkg/tools/packaging/dpackager -e GITHUB_TOKEN=**** -- gh pr edit 18864 --remove-label 'status/ci in progress'
> 13:23:48  error connecting to api.github.com
> 13:23:48  check your internet connection or https://githubstatus.com/
> ```

@yaronkaikov  ?

yaronkaikov commented:
> > > I see "Unit Tests Custom" failed even though it is marked as Pending.
> > 
> > 
> > At least it's not due to failed unit tests. All of them passed. In the logs there is information:
> > ```
> > 13:23:28  + /jenkins/workspace/scylla-master/scylla-ci/scylla-pkg/tools/packaging/dpackager -e GITHUB_TOKEN=**** -- gh pr edit 18864 --remove-label 'status/ci in progress'
> > 13:23:48  error connecting to api.github.com
> > 13:23:48  check your internet connection or https://githubstatus.com/
> > ```
> 
> @yaronkaikov ?

will have a look

yaronkaikov commented:
> 13:23:28  + /jenkins/workspace/scylla-master/scylla-ci/scylla-pkg/tools/packaging/dpackager -e GITHUB_TOKEN=**** -- gh pr edit 18864 --remove-label 'status/ci in progress'
> 13:23:48  error connecting to api.github.com
> 13:23:48  check your internet connection or https://githubstatus.com/

Sent a patch to fix - https://github.com/scylladb/scylla-pkg/pull/4216

paszkow commented:
In v7:

- Addressed the last comment from Botond and therefore the logic in `storage_proxy_::remote` that detects reverse format got even more simplified

denesb commented on service/storage_proxy.cc:
```diff
@@ -726,12 +726,35 @@ class storage_proxy::remote {
+            // Therefore there are only two options:
+            // 1. versions are the same -> legacy format
+            // 2. versions are different -> schema version is equal to reversed table schema version -> native format
+            format_reverse_required = cmd->schema_version == table_schema->version() ? true : false;
```
Why not just directly assign:
```c++
format_reverse_required = cmd->schema_version == table_schema->version();
```
This is perfectly valid C++. The ?: operator makes this more confusing.

paszkow replied:
That's so true. I was just so focused on removing the previous function call that produced this.

denesb reviewed (approved):
Single nitpick, LGTM otherwise.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10373/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10373/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/counter_test
:small_blue_diamond: boost/multishard_mutation_query_test
:small_blue_diamond: boost/mutation_query_test
:small_blue_diamond: boost/mutation_reader_another_test
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/row_cache_test
:small_blue_diamond: boost/sstable_conforms_to_mutation_source_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10373/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10602/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2583/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10373/)
#### Build Details:
- Duration: 4 hr 36 min
- Builder: spider3.cloudius-systems.com

denesb commented:
@paszkow please rebase and resolve the last outstanding comment, so we can get this in.

paszkow commented:
> @paszkow please rebase and resolve the last outstanding comment, so we can get this in.

@denesb Rebased and the nitpick corrected.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10887/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10887/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/counter_test
:small_blue_diamond: boost/multishard_mutation_query_test
:small_blue_diamond: boost/mutation_query_test
:small_blue_diamond: boost/mutation_reader_another_test
:small_blue_diamond: boost/mutation_reader_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: boost/row_cache_test
:small_blue_diamond: boost/sstable_conforms_to_mutation_source_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10887/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11046/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10887/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3021/)
#### Build Details:
- Duration: 4 hr 59 min
- Builder: spider1.cloudius-systems.com
